src/time/Waveform.cpp

```diff
@@ -0,0 +1,154 @@
+    sampleStorageSize = 1;
+  } else {
+    PRECICE_ASSERT(_interpolationOrder >= 0);
+    sampleStorageSize = std::max({_interpolationOrder + 1});
+  }
+  _timeWindowsStorage    = Eigen::MatrixXd::Zero(valuesSize, sampleStorageSize);
+  _numberOfStoredSamples = 1; // the first sample is automatically initialized as zero and stored.
+  _storageIsInitialized  = true;
+  PRECICE_ASSERT(this->sizeOfSampleStorage() == sampleStorageSize);
+  PRECICE_ASSERT(this->valuesSize() == valuesSize);
+}
+
+void Waveform::store(const Eigen::VectorXd &values)
+{
+  PRECICE_ASSERT(_storageIsInitialized);
+  int columnID = 0;
+  this->storeAt(values, columnID);
+}
+
+void Waveform::storeAt(const Eigen::VectorXd values, int columnID)
+{
+  PRECICE_ASSERT(_timeWindowsStorage.cols() > columnID, sizeOfSampleStorage(), columnID);
+  PRECICE_ASSERT(values.size() == this->valuesSize(), values.size(), this->valuesSize());
+  this->_timeWindowsStorage.col(columnID) = values;
+}
+
+Eigen::VectorXd Waveform::sample(double normalizedDt)
+{
+  PRECICE_ASSERT(_storageIsInitialized);
+  PRECICE_ASSERT(normalizedDt >= 0, "Sampling outside of valid range!");
+  PRECICE_ASSERT(normalizedDt <= 1, "Sampling outside of valid range!");
+  if (_interpolationOrder == Time::UNDEFINED_INTERPOLATION_ORDER) {
+    return this->_timeWindowsStorage.col(0);
+  } else {
+    return this->interpolate(normalizedDt);
+  }
+}
+
+void Waveform::moveToNextWindow()
+{
+  PRECICE_ASSERT(_storageIsInitialized);
+  auto initialGuess = _timeWindowsStorage.col(0);                // use value from last window as initial guess for next
+  utils::shiftSetFirst(this->_timeWindowsStorage, initialGuess); // archive old samples and store initial guess
+  if (_numberOfStoredSamples < sizeOfSampleStorage()) {          // together with the initial guess the number of stored samples increases
+    _numberOfStoredSamples++;
+  }
+}
+
+int Waveform::sizeOfSampleStorage()
+{
+  PRECICE_ASSERT(_storageIsInitialized);
+  return _timeWindowsStorage.cols();
+}
+
+Eigen::VectorXd Waveform::getSample(int sampleID)
+{
+  //PRECICE_ASSERT(sampleID < _numberOfStoredSamples);  // @todo use this stricted assertion?
+  PRECICE_ASSERT(sampleID < sizeOfSampleStorage());
+  return _timeWindowsStorage.col(sampleID);
+}
+
+int Waveform::valuesSize()
+{
+  PRECICE_ASSERT(_storageIsInitialized);
+  return _timeWindowsStorage.rows();
+}
+
+/**
+ * @brief Computes which order may be used for interpolation.
+ * 
+ * Order of interpolation is determined by number of stored samples and maximum order defined by the user.
+ * Example: If only two samples are available, the maximum order we may use is 1, even if the user demands order 2.
+ *
+ * @param requestedOrder Order requested by the user.
+ * @param numberOfAvailableSamples Samples available for interpolation.
+ * @return Order that may be used.
+ */
+static int computeUsedOrder(int requestedOrder, int numberOfAvailableSamples)
```

Comment:
Honestly, I don't have it on the radar. Where should this static method go? From my perspective it is static, because it does not depend on the instance of the Waveform, but it also belongs into the class. Or am I using the keyword totally wrong?

Reply:
In general, we try to avoid any static in preCICE. I guess here it's not critical, but would also be easy to avoid.
Static variables can be tricky when we want to use multiple instances of `SolverInterface` simultaneously.
@fsimonis Am I too strict here?

Reply:
I now changed it to non-static.